lib/tlog/tap.c

```diff
@@ -32,6 +32,117 @@
 #include <time.h>
 #include <semaphore.h>
 #include <assert.h>
+#include <limits.h>
+#include <ctype.h>
+#include <string.h>
+#include <sys/types.h>
+#include <pwd.h>
+
+/**
+ * Add an utmp entry for this session.
+ *
+ * @param tap   tlog_tap structure address.
+ * @param path  The filename of the slave pty.
+ *
+ * @return zero, or -1 in case of error with errno set.
+ */
+static int
+tlog_tap_add_utmp_entry(struct tlog_tap *tap, char *path)
+{
+    int ofs;
+    struct timeval tv;
+    char *ses_line;
+    struct utmpx ut;
+    struct utmpx *ut_ptr;
+
+    if ((ses_line = ttyname(tap->tty_fd)) == NULL)
+        return -1;
+
+    if (gettimeofday(&tv, NULL) < 0)
+        return -1;
+
+    if ((tap->ut = calloc(sizeof(*tap->ut), 1)) == NULL)
+        return -1;
+
+    /* Set basic utmp entry data */
+    strncpy(ut.ut_user, getpwuid(getuid())->pw_name, sizeof(ut.ut_user));
+    ut.ut_type = USER_PROCESS;
+    ut.ut_pid = tap->pid;
+
+    /* Strip the leading "/dev/" to get the line name */
+    ofs = strncmp(path, TLOG_TAP_DEVPATH, TLOG_TAP_DEVSIZE) == 0 ? TLOG_TAP_DEVSIZE : 0;
+    memcpy(ut.ut_line, path + ofs, TLOG_TAP_UTMP_LINESIZE - 1);
+
+    /* Copy the trailing pty number to ut_id */
+    ofs = strlen(path) - sizeof(ut.ut_id);
+    if (ofs > 0) {
+        memcpy(ut.ut_id, path + ofs, sizeof(ut.ut_id));
+    }
+
+    /* Set the session start time */
+    ut.ut_tv.tv_sec = tv.tv_sec;
+    ut.ut_tv.tv_usec = tv.tv_usec;
+
+    /* Find the current session utmp entry, if any, and copy the remaining
+     * data to the new entry */
+    ofs = strncmp(ses_line, TLOG_TAP_DEVPATH, TLOG_TAP_DEVSIZE) == 0 ? TLOG_TAP_DEVSIZE : 0;
+    ses_line += ofs;
+    while ((ut_ptr = getutxent()) != NULL &&
+            (ut_ptr->ut_type != USER_PROCESS || strcmp(ut_ptr->ut_line, ses_line)))
+        ;
+    if (ut_ptr) {
+        memcpy(ut.ut_host, ut_ptr->ut_host, sizeof(ut_ptr->ut_host));
+        ut.ut_session = ut_ptr->ut_session;
+        memcpy(ut.ut_addr_v6, ut_ptr->ut_addr_v6, sizeof(ut_ptr->ut_addr_v6));
+    }
+
+    /* Add the new session entry */
+    setutxent();
+    ut_ptr = pututxline(&ut);
+    endutxent();
+
+    if (ut_ptr != NULL) {
+        memcpy(tap->ut, &ut, sizeof(ut));
+        return 0;
+    }
+
+    free(tap->ut);
+    tap->ut = NULL;
+    return -1;
```

Comment:
Looking at the source of other login-related applications, we don't want to fail if pututxline() returns NULL - we can ignore failures.